File: Mavim-apps/Mavim.frontend/projects/mavim-app/src/shared/field/mappers/multidatefield.mapper.ts
import { Field } from '../models/field.model';
import { MultiDateField } from '../models/api';
import { GenericDateFieldMapper } from './abstract';
import { mapToApiFieldType, mapToFieldType } from './fieldvaluetype';

export class MultiDateFieldMapper extends GenericDateFieldMapper<MultiDateField> {
	public get getUrlTypeName(): string {
		return 'multidate';
	}

	public getMappedField(field: MultiDateField): Field {
		return this.mapField(field);
	}

	public getMappedApiField(field: Field): MultiDateField {
		return this.mapApiField(field);
	}

	protected mapApiField(field: Field): MultiDateField {
		const data =
			field.data && field.data.length > 0
				? field.data.map((f) => this.mapToDate(f))
				: [];
		return {
			...field,
			fieldValueType: mapToApiFieldType(field.fieldValueType),
			data
		} as MultiDateField;
	}

	protected mapField(field: MultiDateField): Field {
		const data =
			field.data && field.data.length > 0
				? field.data.map((f) => this.mapToString(f))
				: [];

		return {
			...field,
			fieldValueType: mapToFieldType(field.fieldValueType),
			data
		} as Field;
	}
}

File: Mavim-services/Topic/Mavim.Manager.Api.Topic.Repository/v1/Mappers/MultiDateFieldMapper.cs
using Mavim.Manager.Api.Topic.Repository.Interfaces.v1.enums;
using Mavim.Manager.Api.Topic.Repository.Interfaces.v1.Fields;
using Mavim.Manager.Api.Topic.Repository.v1.Fields;
using Mavim.Manager.Api.Topic.Repository.v1.Mappers.Abstract;
using System;
using System.Linq;

namespace Mavim.Manager.Api.Topic.Repository.v1.Mappers
{
    internal class MultiDateFieldMapper : GenericDateFieldMapper<Model.IMultiValueSimpleField, IMultiDateField>
    {
        /// <summary>
        /// Gets the generic mapped multi date service field.
        /// </summary>
        /// <param name="field">The field.</param>
        /// <returns></returns>
        protected override IMultiDateField GetGenericMappedServiceField(Model.IMultiValueSimpleField field)
        {
            if (field == null) throw new ArgumentNullException(nameof(field));

            return new MultiDateField
            {
                FieldId = field.FieldDefinition.ID,
                FieldName = field.FieldDefinition.Name,
                SetOrder = field.FieldSetDefinition.OrderNumber,
                Order = field.OrderNumber,
                FieldSetId = field.FieldSetDefinition.ID,
                SetName = field.FieldSetDefinition?.Name ?? string.Empty,
                FieldValues = field.FieldValues?.Select(Map),
                FieldValueType = FieldType.MultiDate,
                Required = field.FieldDefinition.Required,
                Readonly = field.ReadOnly
            };
        }

        /// <summary>
        /// Gets the generic mapped multi date repo field.
        /// </summary>
        /// <param name="field">The field.</param>
        /// <param name="simpleField">The simple field.</param>
        /// <returns></returns>
        protected override object[] GetGenericMappedRepoField(IMultiDateField field, Model.ISimpleField simpleField = null)
        {
            if (field == null) return new object[] { null };

            DateTime?[] multiDateField = field.FieldValues.ToArray();

            object[] returnDates = new object[multiDateField.Length];

            for (int i = 0; i < multiDateField.Length; i++)
            {
                if (multiDateField[i] != null)
                {
                    // DateTime is a value type so you can't assign a DateTime[] to an object[] variable.
                    // You'll have to explicitly create an object array and copy the values over.
                    DateTime?[] dateTimeArray = { ((DateTime)multiDateField[i]).ToUniversalTime() };
                    dateTimeArray.CopyTo(returnDates, i);
                }
                else
                {
                    returnDates[i] = null;
                }
            }

            return returnDates;
        }

    }
}
